bpf/cilium-probe-kernel-hz.c

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (C) 2020 Authors of Cilium */
+
+#define _GNU_SOURCE
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sched.h>
+#include <errno.h>
+#include <string.h>
+#include <time.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <sys/resource.h>
+
+struct cpu_jiffies {
+	uint64_t *jiffies;
+	uint32_t cpus;
+};
+
+static const uint64_t kernel_hz[] = { 100, 250, 300, 1000 };
+
+#define abs(x)	({ x < 0 ? -x : x; })
+
+static int pin_to_cpu(int cpu)
+{
+	cpu_set_t set;
+
+	CPU_ZERO(&set);
+	CPU_SET(cpu, &set);
+
+	return sched_setaffinity(getpid(), sizeof(set), &set);
+}
+
+static int fix_priority(void)
+{
+	struct sched_param sp = {
+		.sched_priority = sched_get_priority_max(SCHED_FIFO),
+	};
+
+	return sched_setscheduler(0, SCHED_FIFO, &sp);
+}
+
+static FILE *timer_list_open(void)
+{
+	return fopen("/proc/timer_list", "r");
+}
+
+static void timer_list_close(FILE *fp)
+{
+	fclose(fp);
+}
+
+static int prep_kern_jiffies(struct cpu_jiffies *before,
+			     struct cpu_jiffies *after)
+{
+	uint64_t jiffies = 0;
+	char buff[512];
+	int cpus = 0;
+	FILE *fp;
+
+	fp = timer_list_open();
+	while (fp && fgets(buff, sizeof(buff), fp)) {
+		if (sscanf(buff, "jiffies: %lu\n", &jiffies) == 1)
+			cpus++;
+	}
+	timer_list_close(fp);
+
+	if (!cpus) {
+		fprintf(stderr, "No procfs support?\n");
+		return -EIO;
+	}
+
+	before->cpus = after->cpus = cpus;
+	before->jiffies = calloc(cpus, sizeof(*before->jiffies));
+	after->jiffies  = calloc(cpus, sizeof(*before->jiffies));
```

Comment:
Maybe check `calloc` return values for `NULL` here?

Reply:
They are checked for NULL right afterwards.

Reply:
Time for me to go to bed 🙄